src/Storages/StorageMergeTree.cpp

```diff
@@ -141,6 +141,10 @@ void StorageMergeTree::startup()
 
 void StorageMergeTree::flush()
 {
+    if (flush_called)
```

Comment:
Do I understand correctly that intention of this patch to make `flush()` to be called only once? It seems to me that it still can be run concurrently if e.g. 2 threads execute this if-statement before any write operation.

Reply:
You are right. We need `compare_exchange` here. I was confused by existing erroneous code with `shutdown_called` flag.

Reply:
But is it possible for `flush()` to be run concurrently?

AFAICS it is only possible to flush() be called twice

https://github.com/ClickHouse/ClickHouse/blob/8c9266b24c0bb434fd144ca4dad9744109b542b4/src/Databases/DatabasesCommon.cpp#L262-L270

One from `flush()` and another from `flushAndShutdown()`.
Am I missing something?

Reply:
As I remember there was a race between `DETACH` query and shutdown of `Context` which is called on server shutdown.